### PR TITLE
feat: support declaring `DTOField` via `Annotated`

### PR DESCRIPTION
**Pull request by peterschutt**
E.g.:

```py
class A(Struct):
    a: Annotated[int, dto_field("read_only")]
```

For #2351

### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

-

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

-

### CODE REVIEW

**Comment by provinzkraut** [comment]
@peterschutt this seems to be ready?

**Comment by peterschutt** [comment]
> @peterschutt this seems to be ready?

Yeh close - no tests. I have some related changes to PR, however I'm AFK for the weekend.

**Comment by peterschutt** [comment]
If an argument to `Annotated` is not hashable, e.g., `Annotated[int, {"set"}]`, then `msgspec.inspect.type_info()` throws a `TypeError`.

I created https://github.com/jcrist/msgspec/issues/565 for that, however, for the tests to pass we'd either need to wait for that to make a release, or make `DTOField` hashable. I've chosen to do the latter. Any objections @litestar-org/maintainers?

File: litestar/dto/field.py [hunk 1]
```diff
@@ -47,3 +53,28 @@ def dto_field(mark: Literal["read-only", "private"] | Mark) -> dict[str, DTOFiel
...
+    Supports ``DTOField`` to bet set via ``Annotated`` or via a field info/metadata mapping.
+
+    E.g., ``Annotated[str, DTOField(mark="read-only")]`` or ``info=dto_field(mark="read-only")``.
+
+    If a value is found in ``field_info_mapping``, it is prioritized over the field definition's metadata.
+
+    Args:
+        field_definition: A field definition.
+        field_info_mapping: A field metadata/info attribute mapping, e.g., SQLAlchemy's "info" attribute,
+          or dataclasses "metadata" attribute.
```
Comment (JacobCoffee) [inline litestar/dto/field.py:70]:
```suggestion
        field_info_mapping: A field metadata/info attribute mapping, e.g., SQLAlchemy's ``info`` attribute,
          or dataclasses ``metadata`` attribute.
```

**Comment by JacobCoffee** [comment]
Is that `unsafe_hash` needed for msgpsec?

**Comment by peterschutt** [comment]
> Is that `unsafe_hash` needed for msgpsec?

Yep. Msgspec passes the type into `dict.get()` which calls `_AnnotatedAlias.__hash__()` which in turn passes a tuple of the instance and its metadata to `hash()`. This means that for `Annotated` to be hashable, all of its metadata must be hashable, and `DTOField` isn't.

I ended up going with `unsafe_hash=True` as I figure it would be less imposing than making the type `eq=True, frozen=True`. Having said that though, these things are pretty much designed to be instantiated in the global namespace (i.e., at model definition time), and it would be weird to modify them after the fact (although we do modify them to mark read-only et al. based on the `DTOConfig` configuration). So, I'd prefer to make them truly immutable and safe hash, but not sure if that would be too backward incompatible.

The other option is to block this until msgspec cuts their next release and not modify the hashability of the types at all.

**Comment by peterschutt** [comment]
> The other option is to block this until msgspec cuts their next release and not modify the hashability of the types at all.

FYI: waiting for this.

**Comment by peterschutt** [comment]
msgspec release is out - I'll pick this up again soon

**Comment by provinzkraut** [comment]
@peterschutt Anything blocking this aside from time constraints?

**Comment by peterschutt** [comment]
> @peterschutt Anything blocking this aside from time constraints?

Nothing - except that I keep getting distracted with work closer to the top of the pile.

**Comment by peterschutt** [comment]
Superseded by #3289 3289

**Pull request closed without merging**